lib/reparto.ts

```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
-    ){}
+    constructor(readonly compis_a_asignar: Array<Compi>, readonly tareas_a_asignar: Array<Tarea>) {
+        if (compis_a_asignar.length === 0) {
+            throw new Error('El array de compañeros no puede estar vacío');
+        }
+        this._compis = new Array<Compi>();
+        this.compis_a_asignar.forEach(c => this.addCompi(c));
+
+        this._tareasDisponibles = new Array<Tarea>();
+        this.tareas_a_asignar.forEach(t => this.addTarea(t));
+
+        this._goal = this.calcularGoal(this._compis, this._tareasDisponibles);
+    }
+
+    calcularGoal(compis:Array<Compi>, tareasDisponibles:Array<Tarea>): number{
+        const totalPts = tareasDisponibles.map(tarea => tarea.puntuacion).reduce((total, pts) => total + pts, 0);
+        return totalPts/compis.length;
+    }
+
+    get goal(): number {
+        return this._goal;
+    }
+
+    asignarTareas(): Map<Compi,Tarea[]> {
+        let asignaciones = new Map<Compi, Tarea[]>();
+        this._compis.forEach(compi => asignaciones.set(compi, []));
+        // Ordenar de menor a mayor horas disponibles
+        const compisOrdenados = [...this._compis].sort((a,b) => a.horasDisponibles - b.horasDisponibles);
+        // Ordenar de mayor a menor puntuación
+        const tareasOrdenadas = [...this._tareasDisponibles].sort((a,b) => b.puntuacion - a.puntuacion);
+
+        tareasOrdenadas.forEach(tarea => {
+            const compisDisponibles = compisOrdenados.filter(compi => {
+                const tiempoAsignado = this.calcularTiempoAsignado(compi, asignaciones);
+                return tiempoAsignado + tarea.duracionEstimada <= compi.horasDisponibles;
+            }).filter(compi => {
+                const ptsAsignados = this.calcularPuntuacion(compi, asignaciones);
+                return ptsAsignados < this._goal;
+            });
+
+            if (compisDisponibles.length > 0) {
+                const compiAsignado = compisDisponibles[0];
+                asignaciones.get(compiAsignado)!.push(tarea);
+            }
+        });
+
+        this._compis.forEach(compi => {
+            const ptsAsignados = this.calcularPuntuacion(compi, asignaciones);
+            if (ptsAsignados < this._goal) {
+                this._goal = ptsAsignados;
+            }
+        });
 
-    setGoal(): void{
-        let totalPuntuacion = 0;
-        for (const tarea of this.tareasDisponibles) {
-            totalPuntuacion += tarea.puntuacion;
+        return asignaciones;
+    }
+
+    addTarea(t: Tarea): void {
+        if (t.duracionEstimada <= 0) {
+            throw new Error('La tarea debe tener una duración estimada válida');
+        } 
+        if (t.puntuacion <= 0) {
+            throw new Error('La tarea debe tener una puntuación válida');
         }
-        this.goal = totalPuntuacion/this.compis.length;
+        this._tareasDisponibles.push(t);
+    }
+
+    removeTarea(t: Tarea): void {
+        this._tareasDisponibles.filter(tarea => tarea !== t);
```

Comment:
Esto no hace lo que dice que hace: devuelve las tareas sin una de ellas, pero no la elimina. Conviene que uséis la palabra adecuada para cada una de las variables.
Por otro lado, también le estás pasando la tarea completa en vez de un identificador único de la tarea. Si la tarea necesita un identificador único, útil para almacenarla en la entidad, por ejemplo, eso debería estar en el modelo. Los bucles siempre cuestan tiempo, y los diccionarios ayudan siempre a recuperar la información de forma más eficiente.

Reply:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/f573eb124515b915f7f2c01e3b41549a34bf87e7